platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+                    value: self.smhc.smhc_resp0.read().bits(),
+                    data: match core::mem::replace(&mut guard.data.op, SmhcOp::None) {
+                        SmhcOp::Read { buf, .. } => Some(buf),
+                        _ => None,
+                    },
+                })
+            }
+        };
+        res
+    }
+}
+
+impl IsrData {
+    #[must_use]
+    fn lock<'a>(&'a self, smhc: &'a smhc::RegisterBlock) -> SmhcDataGuard<'a> {
+        // disable interrupts while holding the guard.
+        smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+        let data = unsafe { &mut *(self.data.get()) };
+        SmhcDataGuard { data, smhc }
+    }
+}
+
+impl Drop for SmhcDataGuard<'_> {
+    fn drop(&mut self) {
+        self.smhc.smhc_ctrl.modify(|_, w| w.ine_enb().enable());
+    }
+}
+
+impl SmhcDataGuard<'_> {
+    async fn wait_for_irq(&mut self) {
+        let mut waiting = false;
+        futures::future::poll_fn(|cx| {
+            if waiting {
+                self.smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+                return Poll::Ready(());
+            }
+
+            self.data.waker = Some(cx.waker().clone());
+            waiting = true;
+            self.smhc.smhc_ctrl.modify(|_, w| w.ine_enb().enable());
+
+            Poll::Pending
+        })
+        .await;
+    }
+}
+
+impl Deref for SmhcDataGuard<'_> {
+    type Target = SmhcData;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.data
+    }
+}
+
+impl DerefMut for SmhcDataGuard<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.data
+    }
+}
+
+impl SmhcData {
+    /// Check for error bits in the interrupt status registers.
+    fn error_status(smhc: &smhc::RegisterBlock) -> Option<ErrorKind> {
+        let mint = smhc.smhc_mintsts.read();
+        let idst = smhc.smhc_idst.read();
+
+        if mint.m_dsto_vsd_int().bit_is_set() {
+            Some(ErrorKind::DataStarvationTimeout)
+        } else if mint.m_fu_fo_int().bit_is_set() {
+            Some(ErrorKind::FifoUnderrunOverflow)
+        } else if mint.m_cb_iw_int().bit_is_set() {
+            Some(ErrorKind::CommandBusyIllegalWrite)
+        } else if mint.m_rto_back_int().bit_is_set() {
+            Some(ErrorKind::ResponseTimeout)
+        } else if mint.m_rce_int().bit_is_set() {
+            Some(ErrorKind::ResponseCrc)
+        } else if mint.m_re_int().bit_is_set() {
+            Some(ErrorKind::Response)
+        } else if mint.m_dto_bds_int().bit_is_set() {
+            Some(ErrorKind::DataTimeout)
+        } else if mint.m_dce_int().bit_is_set() {
+            Some(ErrorKind::DataCrc)
+        } else if mint.m_dse_bc_int().bit_is_set() {
+            Some(ErrorKind::DataStart)
+        } else if mint.m_dee_int().bit_is_set() {
+            Some(ErrorKind::DataEnd)
+        } else if idst.des_unavl_int().bit_is_set() || idst.fatal_berr_int().bit_is_set() {
+            Some(ErrorKind::Dma)
+        } else {
+            None
+        }
```

Comment:
nit, take it or leave it: could potentially be a few lines shorter if this was, like, 
```rust
match () {
   _ if mint.whatever_bit().bit_is_set() => Some(ErrorKind::Whatever),
   // ...
}
```
although it's not a huge deal either way.